src/v/resource_mgmt/cpu_profiler.cc

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "resource_mgmt/cpu_profiler.h"
+
+#include "random/generators.h"
+#include "resource_mgmt/logger.h"
+#include "ssx/future-util.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/util/later.hh>
+
+#include <iterator>
+
+namespace resources {
+
+cpu_profiler::cpu_profiler(
+  config::binding<bool>&& enabled,
+  config::binding<std::chrono::milliseconds>&& sample_rate)
+  : _query_timer([this] { poll_samples(); })
+  , _enabled(std::move(enabled))
+  , _sample_rate(std::move(sample_rate)) {
+    _enabled.watch([this] { on_enabled_change(); });
+    _sample_rate.watch([this] { on_sample_rate_change(); });
+}
+
+ss::future<> cpu_profiler::start() {
+    seastar::engine().set_cpu_profiler_period(_sample_rate());
+    seastar::engine().set_cpu_profiler_enabled(_enabled());
+
+    if (_enabled()) {
+        // Arm the timer to fire whenever the profiler collects
+        // the maximum number of traces it can retain.
+        _query_timer.arm_periodic(ss::max_number_of_traces * _sample_rate());
+    }
+
+    return ss::now();
+}
+
+ss::future<> cpu_profiler::stop() {
+    _query_timer.cancel();
+    co_await _gate.close();
+}
+
+void cpu_profiler::poll_samples() { do_poll_samples(); }
```

Comment:
what is the point of this method?

Reply:
No point anymore. Originally `do_poll_samples` was a coroutine so `poll_samples` grabbed a gate before executing it in the background. Will inline the body of `do_poll_samples` into `poll_samples` now though to remove the unneeded indirection.